components/sharing-dialog/src/helpers/helpers.js

```diff
@@ -0,0 +1,249 @@
+import {
+    ACCESS_VIEW_ONLY,
+    ACCESS_NONE,
+    ACCESS_VIEW_AND_EDIT,
+    SHARE_TARGET_EXTERNAL,
+    SHARE_TARGET_PUBLIC,
+} from '../sharing-constants.js'
+
+/**
+ * Returns a function, that, as long as it continues to be invoked, will not be triggered. The
+ * function will be called after it stops being called for N milliseconds. If `immediate` is
+ * passed, trigger the function on the leading edge, instead of the trailing.
+ */
+
+export const debounce = (func, wait, immediate) => {
+    let timeout
+
+    return (...args) => {
+        const context = this
+
+        const later = () => {
+            timeout = null
+
+            if (!immediate) {
+                func.apply(context, args)
+            }
+        }
+
+        const callNow = immediate && !timeout
+
+        clearTimeout(timeout)
+        timeout = setTimeout(later, wait)
+
+        if (callNow) {
+            func.apply(context, args)
+        }
+    }
+}
+
+/**
+ * Name conversion
+ */
+
+export const nameToInitials = (name) => {
+    if (!name) {
+        return ''
+    }
+
+    const parts = name.split(' ')
+    const firstNameInitial = parts.length > 0 ? parts.shift().charAt(0) : ''
+    const lastNameInitial = parts.length > 0 ? parts.pop().charAt(0) : ''
+
+    return `${firstNameInitial}${lastNameInitial}`
+}
+
+/**
+ * Access and constant conversion
+ */
+
+export const convertAccessToConstant = (access) => {
+    if (typeof access === 'boolean') {
+        return access ? ACCESS_VIEW_ONLY : ACCESS_NONE
+    }
+
+    if (/^rw/.test(access)) {
+        return ACCESS_VIEW_AND_EDIT
+    } else if (/^r-/.test(access)) {
+        return ACCESS_VIEW_ONLY
+    } else {
+        return ACCESS_NONE
+    }
+}
+
+export const convertConstantToAccess = (constant, useBoolean) => {
+    switch (constant) {
+        case ACCESS_NONE:
+            return useBoolean ? false : '--------'
+        case ACCESS_VIEW_ONLY:
+            return useBoolean ? true : 'r-------'
+        case ACCESS_VIEW_AND_EDIT:
+            return useBoolean ? true : 'rw------'
+        default:
+            return useBoolean ? false : '--------'
+    }
+}
+
+/**
+ * Mapping users and groups to style used internally
+ */
+
+export const mapUsers = ({ id, name, access }) => ({
+    id,
+    name,
+    access: convertAccessToConstant(access),
+})
+
+export const mapGroups = ({ id, name, access }) => ({
+    id,
+    name,
+    access: convertAccessToConstant(access),
+})
+
+/**
+ * Helper to check whether to allow removing for the selected target
+ */
+
+export const isRemovableTarget = (target) => {
+    const permanentTargets = [SHARE_TARGET_EXTERNAL, SHARE_TARGET_PUBLIC]
+
+    // Do not allow removal of permanent targets
+    return !permanentTargets.includes(target)
+}
+
+/**
+ * Mutation handlers
+ */
+
+export const createOnChange =
```

Comment:
Not a big fan of a function doing two things at the same time (creating a payload & performing a mutate), but I suppose that's highly subjective. I guess my preference would be to not pass the mutate function but instead just pass the created payload to the mutate function

Reply:
Changing this would also make it easier to reason about the arguments that are supplied to this function:

```js
onAdd={({ type, id, access }) => {
  const data = createOnAddPayload({ type, id, access, object })
  mutate({ data })
}}
```

Reply:
Yeah I don't like this section either. It's also a bit annoying that here type and id are shadowing the sharing-dialog type and id props, but they're actually referring to something completely different. I'll see if I can improve this.

Reply:
Good suggestion, I like this pattern a lot better. Changed it in b3cade77b1854ee386f329420559f2f5faec51bc, and I've also updated mutate here: 828ef46f003c59fce70234f0fb365929bba545d2, so that it picks up on prop changes in the root component (otherwise it'd keep using the initial props and update the wrong object).

Reply:
As you can see in the last commit, the type and id params in the handlers shadow the global type and id (which is not ideal). I've updated the naming only for the root component. But we could propagate those names downwards if you think that's clearer.

Reply:
I'll resolve it, but let me know if you think anyone still needs to be addressed.